src/core/server/http/base_path_proxy_server.ts

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { RouteOptions, Server } from 'hapi-latest';
+import { Agent as HttpsAgent, ServerOptions as TlsOptions } from 'https';
+import { sample } from 'lodash';
+import { ByteSizeValue } from '../config/schema';
+import { DevConfig } from '../dev';
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+const alphabet = 'abcdefghijklmnopqrztuvwxyz'.split('');
+
+export interface BasePathProxyServerOptions {
+  httpConfig: HttpConfig;
+  devConfig: DevConfig;
+  isKibanaPath: (path: string) => boolean;
+  blockUntil?: () => Promise<void>;
+}
+
+export class BasePathProxyServer {
+  private server?: Server;
+  private httpsAgent?: HttpsAgent;
+
+  get basePath() {
+    return this.options.httpConfig.basePath;
+  }
+
+  get targetPort() {
+    return this.options.devConfig.basePathProxyTargetPort;
+  }
+
+  constructor(
+    private readonly log: Logger,
+    private readonly options: BasePathProxyServerOptions
+  ) {
+    const ONE_GIGABYTE = 1024 * 1024 * 1024;
+    options.httpConfig.maxPayload = new ByteSizeValue(ONE_GIGABYTE);
+
+    if (!options.httpConfig.basePath) {
```

Comment:
can we be more clear about this, is it false, is it undefined?

Reply:
nevermind, this is the actual string for the basePath. the user would have provided it in the config, if it's not provided, then we set it to a random 3 character string.

Reply:
Yeah, also we get `basePath` from the old platform and it never provides it as `undefined` or `null`, if it's not defined - it will always be empty string (https://github.com/elastic/kibana/blob/8f206db2c50349fa63a10fd6a8399a924ba0ee1c/src/server/config/schema.js#L74).